url1d5e9d990f_Telsis_jOCP_tgz-pJ8-com_telsis_jocp_sampleApp_BasicOCPAppJ8/src/com/telsis/jocp/OCPLinkManager.java

```diff
@@ -0,0 +1,2057 @@
+        /**
+         * The write selector for the socket.
+         */
+        private Selector       socketSelector;
+        /**
+         * The transmit buffer.
+         */
+        private ByteBuffer     txBuffer;
+        /**
+         * A flag to detect when a heartbeat should be sent.
+         */
+        private boolean        needHeartbeat;
+        /**
+         * The timestamp of the last activity.
+         */
+        private long           lastActivity;
+        /**
+         * The timestamp of the last heartbeat transmitted.
+         */
+        private long           lastHeartbeat;
+
+        /**
+         * Creates a new stopped TxThread.
+         */
+        public TxThread() {
+            super();
+
+            shutdownThread = false;
+            myThread = null;
+        }
+
+        @Override
+        public void run() {
+            log.debug("TxThread: thread starting");
+            watchdog.start();
+            myThread = Thread.currentThread();
+            boolean moreToDo = false;
+
+            lastActivity = System.currentTimeMillis();
+            lastHeartbeat = lastActivity;
+
+            try {
+                socketSelector = Selector.open();
+                channel.register(socketSelector, SelectionKey.OP_WRITE);
+                currentState = TxThreadStates.UNINITIALISED;
+
+                while (true) {
+                    watchdog.pat();
+                    moreToDo = internalRun();
+
+                    if (shutdownThread) {
+                        break;
+                    }
+
+                    long now = System.currentTimeMillis();
+                    if (now - lastActivity > inactiveHeartbeatInterval
+                            || now - lastHeartbeat > fixedHeartbeatInterval) {
+                        needHeartbeat = true;
+                        log.debug("TxThread: requested heartbeat");
+                    }
+
+                    if (!moreToDo) {
+                        socketSelector.select(1 * MILLISECONDS);
+                        socketSelector.selectedKeys().clear();
+                    }
+                }
+            } catch (IOException e) {
+                log.error("TxThread: exception in thread", e);
+            } finally {
+                try {
+                    socketSelector.close();
+                } catch (IOException e) { // CSIGNORE: EmptyBlock
+                    // The thread is exiting anyway, so no need to do anything
+                }
+            }
+            watchdog.stop();
+            log.debug("TxThread: thread exiting");
+        }
+
+        /**
+         * Shut down this thread.
+         */
+        public void shutdownThread() {
+            shutdownThread = true;
+            if (myThread != null) {
+                while (myThread.isAlive()) {
+                    try {
+                        myThread.join();
+                    } catch (InterruptedException e) { // CSIGNORE: EmptyBlock
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)